packages/hooks/use-archive-infinite-scroll/index.tsx

```diff
@@ -0,0 +1,231 @@
+    if (!supported) return children;
+
+    const container = css`
+      position: relative;
+    `;
+
+    const fetcher = css`
+      position: absolute;
+      width: 100%;
+      bottom: 0;
+    `;
+
+    return (
+      <div css={container} ref={routeRef} className={className}>
+        {children}
+        {!hasReachedLimit && <div css={fetcher} ref={fetchRef} />}
+      </div>
+    );
+  };
+
+  return connect(Wrapper, { injectProps: false });
+};
+
+/**
+ * A memoized {@link Wrapper} to generate Wrapper components only once.
+ *
+ * @param link - The link for the page that the Wrapper belongs to.
+ * @param options - The intersection observer options for fetching and routing.
+ *
+ * @returns A React component that should be used to wrap the page.
+ */
+const MemoizedWrapper = memoize(
+  (link) => link,
+  (
+    link: string,
+    options: {
+      /**
+       * The intersection observer options for fetching.
+       */
+      fetchInViewOptions: IntersectionOptions;
+
+      /**
+       * The intersection observer options for routing.
+       */
+      routeInViewOptions: IntersectionOptions;
+    }
+  ) =>
+    wrapperGenerator({
+      link,
+      fetchInViewOptions: options.fetchInViewOptions,
+      routeInViewOptions: options.routeInViewOptions,
+    })
+);
+
+/**
+ * A hook used to add infinite scroll to any Frontity archive.
+ *
+ * @param options - Options for the hook, like the number of pages it should
+ * load autoamtically or if it's active or not. Defined in {@link
+ * UseArchiveInfiniteScroll}.
+ *
+ * @returns - An array of pages and other useful booleans. Defined in {@link
+ * UseArchiveInfiniteScroll}.
+ */
+const useArchiveInfiniteScroll = (
+  options: UseArchiveInfiniteScrollOptions = {}
+): UseArchiveInfiniteScrollOutput => {
+  const defaultOptions = { active: true };
+  options = { ...defaultOptions, ...options };
+
+  const { state, actions } = useConnect<Source & Router>();
+  const { infiniteScroll } = state.router.state as InfiniteScrollRouterState;
+
+  // Values from/for browser state.
+  const links: string[] = infiniteScroll?.links
+    ? [...infiniteScroll?.links]
+    : [state.router.link];
+  const limit: number = infiniteScroll?.limit || options.limit;
+
+  // Initialize/update browser state.
+  useEffect(() => {
+    if (!options.active) return;
+
+    actions.router.updateState({
+      ...state.router.state,
+      infiniteScroll: {
+        links,
+        limit,
+        ...infiniteScroll,
+      },
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [options.active]);
+
+  // Aliases to needed state.
+  const last = state.source.get(links[links.length - 1]);
+  const next =
+    isArchive(last) && last.next ? state.source.get(last.next) : null;
+
+  // Infinite scroll booleans.
+  const hasReachedLimit = !!limit && links.length >= limit;
+  const thereIsNext = isArchive(last) && last.next;
+  const isFetching = last.isFetching || !!next?.isFetching;
```

Comment:
I have a question here: by using `!!next` you are turning the `next` data object into a boolean.

So this code `!!next?.isFetching` means: 
1. Turn the `next` data object in a boolean.
2. Check if it is truthy. As it is a boolean, that can only be `true`.
3. If it is `true`, check its `isFetching` prop.

So if I am not mistaken, you are doing `true.isFetching`, which doesn't fail, but always returns `undefined`.

Reply:
As far as I know, the Optional Chaining operator (`.?`) takes precedence over the Logical NOT (`!`) operator, same as the Member Access operator (`.`).

Here is a [table from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#table) with all the JS operators listed by precedence, I guess TypeScript maintains the same order.

So, actually, `!!next?.isFetching` means `!! (next?.isFetching)`. :blush:

Reply:
I just did a quick test on the TypeScript Playground and, yeah, this is the way the code above is compiled:

```js
!!(next === null || next === void 0 ? void 0 : next.isFetching);
```

Reply:
Ok, thanks for the clarification.

By the way, what is the benefit of using `!!next?.isFetching` over `next?.isFetching`? The type?

Reply:
Yest, it's just a way to ensure that `isFetching` holds a boolean value. Also, that's the value the hook returns, so I think it makes sense.

Reply:
Perfect. Thanks for the explanation David 👍